File: src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.java
package com.autosale.shop.repository.impl;

import com.autosale.shop.model.PaginationRequest;
import com.autosale.shop.model.Product;
import com.autosale.shop.repository.ProductRepository;
import lombok.RequiredArgsConstructor;
import org.jooq.Condition;
import org.jooq.DSLContext;
import org.jooq.exception.DataAccessException;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Isolation;
import org.springframework.transaction.annotation.Transactional;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import static structure.tables.Product.PRODUCT;

@Repository
@RequiredArgsConstructor
public class ProductRepositoryImpl implements ProductRepository {
    private final DSLContext dsl;

    @Override
    @Transactional(isolation = Isolation.READ_UNCOMMITTED)
    // читання даних не відбувається а номер айді заповює сама бд, тому можна поставити найнижчий рівень ізоляції
    public Optional<Integer> save(Product product) {
        return dsl.insertInto(PRODUCT)
                .set(dsl.newRecord(PRODUCT, product))
                .returningResult(PRODUCT.ID)
                .fetchOptionalInto(Integer.class);

    }

    @Override
    public Optional<Product> findById(int id) {
        return dsl.selectFrom(PRODUCT)
                .where(PRODUCT.ID.eq(id))
                .fetchOptionalInto(Product.class);
    }

    @Override
    public void update(Product product) {
        int deleted = dsl.deleteFrom(PRODUCT)
                .where(PRODUCT.ID.eq(product.getId()))
                .execute();
        if (deleted == 0) {
            throw new DataAccessException("No record with id " + product.getId() + " found for deletion.");
        }
        dsl.insertInto(PRODUCT)
                .set(dsl.newRecord(PRODUCT, product))
                .execute();
    }

    @Override
    public int deleteById(int id) {
        return dsl.deleteFrom(PRODUCT)
                .where(PRODUCT.ID.eq(id))
                .execute();
    }

    @Override
    public int deleteByIdInArray(List<Integer> ids) {
        return dsl.deleteFrom(PRODUCT)
                .where(PRODUCT.ID.in(ids))
                .execute();
    }

    @Override
    public List<Product> findAllByUserId(int userId) {
        return findAllWithConditions(List.of(PRODUCT.SELLER_ID.eq(userId)));
    }

    @Override
    @Transactional(isolation = Isolation.READ_COMMITTED) // йде зчитування актуальних даних на момент запиту
    public List<Product> findAllByStatus(PaginationRequest pageRequest, String status) {
        if (status != null) {
            return findAllWithConditionsPageable(pageRequest, List.of(PRODUCT.STATUS.eq(status)));
        } else {
            return findAllWithConditionsPageable(pageRequest, Collections.emptyList());
        }
    }

    @Override
    public int countAllByStatus(PaginationRequest pageRequest, String status) {
        if (status != null) {
            return countAllWithConditions(List.of(PRODUCT.STATUS.eq(status)));
        } else {
            return countAllWithConditions(Collections.emptyList());
        }
    }

    @Override
    public void saveAllIgnoreExistence(List<Product> products) {

        dsl.batchStore(products.stream()
                .map(p -> dsl.newRecord(PRODUCT, p))
                .collect(Collectors.toUnmodifiableList()))
                .execute();
    }

    private List<Product> findAllWithConditions(List<Condition> conditions) {
        return dsl.selectFrom(PRODUCT)
                .where(conditions)
                .fetchInto(Product.class);
    }

    private List<Product> findAllWithConditionsPageable(PaginationRequest paginationRequest, List<Condition> conditions) {
        return dsl.selectFrom(PRODUCT)
                .where(conditions)
                .offset((paginationRequest.getCurrentPage() - 1) * paginationRequest.getPageSize()) // -1 because we start pages from 1
                .limit(paginationRequest.getPageSize())
                .fetchInto(Product.class);
    }

    private int countAllWithConditions(List<Condition> conditions) {
        return dsl.fetchCount(PRODUCT, conditions);
    }

}

File: src/main/java/com/autosale/shop/repository/impl/ProductRepositoryImpl.kt
package com.autosale.shop.repository.impl

import com.autosale.shop.model.PaginationRequest
import com.autosale.shop.model.Product
import com.autosale.shop.repository.ProductRepository
import org.jooq.Condition
import org.jooq.DSLContext
import org.springframework.stereotype.Repository

@Repository
class ProductRepositoryImpl(private val dsl: DSLContext) : ProductRepository {
    
    override fun save(product: Product): Int {
        return dsl.insertInto(structure.tables.Product.PRODUCT)
            .set(dsl.newRecord(structure.tables.Product.PRODUCT, product))
            .returningResult(structure.tables.Product.PRODUCT.ID)
            .fetchOneInto(Int::class.java)!!
    }

    override fun findById(id: Int): Product? {
        return dsl.selectFrom(structure.tables.Product.PRODUCT)
            .where(structure.tables.Product.PRODUCT.ID.eq(id))
            .fetchOneInto(Product::class.java)
    }

    override fun update(product: Product) {
        dsl.newRecord(structure.tables.Product.PRODUCT, product)
            .update()
    }

    override fun deleteById(id: Int): Int {
        return dsl.deleteFrom(structure.tables.Product.PRODUCT)
            .where(structure.tables.Product.PRODUCT.ID.eq(id))
            .execute()
    }

    override fun deleteByIdInArray(ids: List<Int>): Int {
        return dsl.deleteFrom(structure.tables.Product.PRODUCT)
            .where(structure.tables.Product.PRODUCT.ID.`in`(ids))
            .execute()
    }

    override fun findAllByUserId(userId: Int): List<Product> {
        return findAllWithConditions(listOf(structure.tables.Product.PRODUCT.SELLER_ID.eq(userId)))
    }

    override fun findAllByStatus(pageRequest: PaginationRequest, status: String?): List<Product> {
        return if (status != null) {
            findAllWithConditionsPageable(
                pageRequest,
                listOf(structure.tables.Product.PRODUCT.STATUS.eq(status))
            )
        } else {
            findAllWithConditionsPageable(pageRequest, emptyList())
        }
    }

    override fun countAllByStatus(pageRequest: PaginationRequest, status: String?): Int {
        return if (status != null) {
            countAllWithConditions(listOf(structure.tables.Product.PRODUCT.STATUS.eq(status)))
        } else {
            countAllWithConditions(emptyList())
        }
    }

    override fun saveAllIgnoreExistence(products: List<Product>) {
        dsl.batchStore(products.map { p: Product -> dsl.newRecord(structure.tables.Product.PRODUCT, p) })
            .execute()
    }

    private fun findAllWithConditions(conditions: List<Condition>): List<Product> {
        return dsl.selectFrom(structure.tables.Product.PRODUCT)
            .where(conditions)
            .fetchInto(Product::class.java)
    }

    private fun findAllWithConditionsPageable(
        paginationRequest: PaginationRequest,
        conditions: List<Condition>
    ): List<Product> {
        return dsl.selectFrom(structure.tables.Product.PRODUCT)
            .where(conditions)
            .offset((paginationRequest.currentPage - 1) * paginationRequest.pageSize) // -1 because we start pages from 1
            .limit(paginationRequest.pageSize)
            .fetchInto(Product::class.java)
    }

    private fun countAllWithConditions(conditions: List<Condition>): Int {
        return dsl.fetchCount(structure.tables.Product.PRODUCT, conditions)
    }
}
